testing/trino-testing-services/src/main/java/io/trino/jvm/Threads.java

```diff
@@ -0,0 +1,103 @@
+
+    /**
+     * Returns representation of {@link ThreadInfo}. Similar to {@link ThreadInfo#toString()} but
+     * without a hard-coded limit on number or returned stack frames.
+     */
+    public static String fullToString(ThreadInfo thread)
+    {
+        StringBuilder sb = new StringBuilder();
+
+        sb.append(format(
+                "\"%s\"%s prio=%s Id=%s %s%s%s%s%s\n",
+                thread.getThreadName(),
+                thread.isDaemon() ? " daemon" : "",
+                thread.getPriority(),
+                thread.getThreadId(),
+                thread.getThreadState(),
+                thread.getLockName() != null ? " on " + thread.getLockName() : "",
+                thread.getLockOwnerName() != null ? " owned by " + thread.getLockOwnerName() : "",
+                thread.isSuspended() ? " (suspended)" : "",
+                thread.isInNative() ? " (in native)" : ""));
+
+        Iterator<StackTraceElement> stackTrace = Arrays.asList(thread.getStackTrace()).iterator();
+
+        Multimap<Integer, MonitorInfo> lockedMonitors = Arrays.stream(thread.getLockedMonitors())
+                .collect(Multimaps.toMultimap(MonitorInfo::getLockedStackDepth, Function.identity(), ArrayListMultimap::create));
+
+        int depth = 0;
+        if (stackTrace.hasNext()) {
+            StackTraceElement first = stackTrace.next();
+            sb.append(format("\tat %s\n", first));
+            LockInfo lockInfo = thread.getLockInfo();
+            if (lockInfo != null) {
+                String lockVerb;
+                switch (thread.getThreadState()) {
+                    case BLOCKED:
+                        lockVerb = "blocked";
+                        break;
+                    case WAITING:
+                    case TIMED_WAITING:
+                        lockVerb = "waiting";
+                        break;
+                    default:
+                        lockVerb = "(unexpected lock info)";
+                }
+                sb.append(format("\t-  %s on %s\n", lockVerb, lockInfo));
+            }
+
+            lockedMonitors.get(depth).forEach(monitor -> sb.append(format("\t-  locked %s\n", monitor)));
+        }
+
+        stackTrace.forEachRemaining(element -> sb.append(format("\tat %s\n", element)));
+        while (stackTrace.hasNext()) {
+            StackTraceElement element = stackTrace.next();
+            depth++;
+            sb.append(format("\tat %s\n", element));
+            lockedMonitors.get(depth).forEach(monitor -> sb.append(format("\t-  locked %s\n", monitor)));
+        }
+
+        LockInfo[] lockedSynchronizers = thread.getLockedSynchronizers();
+        if (lockedSynchronizers.length != 0) {
+            sb.append("\n");
+            sb.append("\tlocked synchronizers:\n");
+            for (LockInfo locked : lockedSynchronizers) {
+                sb.append(format("\t- %s\n", locked));
+            }
+        }
+
+        sb.append("\n");
+
+        return sb.toString();
```

Comment:
Any reasonable limits that should be in place? It would be worse if the string cannot be built and we end up without a trace at all.

Reply:
`StringBuilder` uses a single byte array for the buffer, so it would have to limit during the append. But even with 10,000 threads at 10kB each, that's only 100MB, so probably fine. Worst case would be a huge number of threads all with `StackOverflowException` (maximum stack size), but that doesn't seem possible in practice.